src/generate_conf.rs

```diff
@@ -76,19 +77,39 @@ fn generate_config(
 ) -> Result<(Vec<Interface>, NetworkConfig), anyhow::Error> {
     let network_state = NetworkState::new_from_yaml(&data)?;
 
-    let interfaces = extract_interfaces(&network_state);
-    validate_interfaces(&interfaces, require_mac_addresses)?;
-
     let config = network_state
         .gen_conf()?
         .get("NetworkManager")
         .ok_or_else(|| anyhow!("Invalid NM configuration"))?
         .to_owned();
 
+    let interfaces = extract_interfaces(&network_state, &config);
+    validate_interfaces(&interfaces, require_mac_addresses)?;
```

Comment:
I'd suggest to revert this as this is the validation step.

`extract_interfaces` should only extract the interfaces from the populated network state. Once we have that we can validate that the input is expected. Once that's done, we generate the configuration and then we can extract the connection IDs and merge them in the final result for this function. This would separate the functionality and simplify the flow. WDYT?

Reply:
You would suggest something like:

```rust
    let interfaces = extract_interfaces(&network_state);
    validate_interfaces(&interfaces, require_mac_addresses)?;

    let config = network_state
        .gen_conf()?
        .get("NetworkManager")
        .ok_or_else(|| anyhow!("Invalid NM configuration"))?
        .to_owned();

    populate_connection_ids(&network_state, &interfaces, &config);
```

Reply:
Yes, exactly.